docs/contributing/unit-test-guideline.md

```diff
@@ -0,0 +1,193 @@
+1. Test behaviors and results, not implementation
+	- ensures that tests only fail when there is an actual effect and not due to internal changes.
+1. Test one thing at a time
+	- Each test should have a clear, concise, singular objective.
+1. Make tests readable and understandable
+	- Hold test code to a similar standard as production code.
+1. Make tests deterministic
+	- A test should pass all the time or fail all the time until fixed.
+1. Make tests independent and self-sufficient
+	- Setup, execution, and verification steps in a given test should not depend on running other tests before it. To keep unit tests simple, fast running, and easy to debug, it may be necessary to isolate the class under test.
+1. Repeat yourself when necessary
+	- It is okay to violate the 'do not repeat yourself' principle if it makes tests simpler and easier to read.
+1. Measure code coverage but focus on test coverage
+	- Do not simply to archive code coverage.
+
+### Test case
+
+#### Basic
+
+It is not perfect, but we should try to cover all codes at first.
+To cover all code, the basic way is the creating test cases to complete all branches in the target unit.
+
+Let's see the below function.
+```go
+func calcSum(val ...int32) (sum int32) {
+	if len(val) == 0 {
+		return
+	} else {
+		for _, v := range val {
+			sum += v
+		}
+		return
+	}
+}
+```
+
+When the above function is given, we should create 2 test for achieving 100% code coverage at least.
+
+- When val is not given
+  - In other words, the default value is given
+- When val(int32) is given
+  - `var val int32 = 1`
+  
+It seems enough for the given function, but we have to take care of some test cases that are remaining for improving test coverage.
+In this case, there is one test case is remaining.
+
+- When val([]int32) is given
+  - `var val []int32 = {1, 2}`
+  
+That is the focus on test coverage.
+
+Therefore, we should consider all cases for improving test coverage of the target unit.
+To improve test coverage, the basic but critical thinking way is thinking about input patterns.
+It is not only a single input, but also multi inputs.
+
+If a function or method requires multi-input, we should try to create many test cases.
+
+(Note: The below `calcSum()` is diffrent function from `calsSum(val ...int32)` which we mention before.)
+
+```go
+func calcAverageDiff(val1 []int32, val2 []int32) (diff float64) {
+	var ave1, ave2 float64
+	if len(val1) == 0 && len(val2) == 0 {
+		return
+	}
+	if len(val1) != 0 {
+		ave1 = float64(calcSum(val1)) / float64(len(val1))
+	}
+	if len(val2) != 0 {
+		ave2 = float64(calcSum(val2)) / float64(len(val2))
+	}
+
+	diff = math.Abs(ave1 - ave2)
+	return
+}
+```
+
+When `calcAverageDiff()` is given, the test patterns are below:
+
+| len(val1) | len(val2) | option |
+|:-----:|:-----:|:-----:|
+| 0  | 0  | - |
+| 0  | >0 | - |
+| >0 | 0  | - |
+| >0 | >0 | ave1 > ave2 |
+| >0 | >0 | ave1 < ave2 |
```

Comment:
sane as before, how about boundary values, overflow, and negative value?

Reply:
I think it is not needed.
Both `ave1` and `ave2` are calculated variable, so we do not have to test all of you mentioned.
At least, we have to test overflows.
btw, what did you mean boundary values?

Reply:
boundary values means the Max/Min values. I think you have covered it already.

but do we need to cover the overflow cases here? as `calcSum` will overflow the value

Reply:
I could not reproduce such overflow matter.
could you show me in the playground?

Reply:
for example this.
https://play.golang.org/p/mIiLeW6Qf9j
🤔

Reply:
thx. I got it.